482p3/lib/libc/gen/fmtmsg.3
.\"	$NetBSD: fmtmsg.3,v 1.7 2011/04/11 05:59:11 jruoho Exp $
.\"
.\" Copyright (c) 1999 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Klaus Klein.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd April 11, 2011
.Dt FMTMSG 3
.Os
.Sh NAME
.Nm fmtmsg
.Nd format and display a message
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In fmtmsg.h
.Ft int
.Fn fmtmsg "long classification" "const char *label" "int severity" "const char *text" "const char *action" "const char *tag"
.Sh DESCRIPTION
The
.Fn fmtmsg
function can be used to display messages in the specified format.
Messages may be written either to standard error, to the console, or both.
.Pp
A formatted message consists of up to five components specified in
.Fa label ,
.Fa severity ,
.Fa text ,
.Fa action
and
.Fa tag .
Further information such as the origin of the message, the recoverability
from the condition causing the message and where to display the message
is specified in
.Fa classification .
.Ss Classification
The
.Fa classification
argument consists of a major classification and several sub-classifications.
It has no effect on the content of the message displayed.
With the exception of the display sub-classification, only a single identifier
may be specified for each (sub-)classification.
The following classifications
are available:
.Bl -tag -width "XXX"
.It Em Major classifications
The source of the condition.
Available identifiers are:
.Dv MM_HARD
(hardware),
.Dv MM_SOFT
(software), and
.Dv MM_FIRM
(firmware).
.It Em Message source sub-classifications
The type of software detecting the condition.
Available identifiers are:
.Dv MM_APPL
(application),
.Dv MM_UTIL
(utility), and
.Dv MM_OPSYS
(operating system).
.It Em Display sub-classifications
The displays the formatted messages is to be written to.
Available identifiers are:
.Dv MM_PRINT
(standard error stream) and
.Dv MM_CONSOLE
(system console).
.It Em Status sub-classifications
The capability of the calling software to recover from the condition.
Available identifiers are:
.Dv MM_RECOVER
(recoverable) and
.Dv MM_NRECOV
(non-recoverable).
.El
.Pp
If no
.Fa classification
is to be supplied,
.Dv MM_NULLMC
must be specified.
.Ss Label
The
.Fa label
argument identifies the source of the message.
It consists of two fields separated by a colon (:).
The first field is up to 10 characters, the second is up to 14 characters.
.Pp
If no
.Fa label
is to be supplied,
.Dv MM_NULLLBL
must be specified.
.Ss Severity
The seriousness of the condition causing the message.
The following
.Fa severity
levels are available:
.Bl -tag -width MM_WARNING -offset indent
.It Dv MM_HALT
The software has encountered a severe fault and is halting.
.It Dv MM_ERROR
The software has encountered a fault.
.It Dv MM_WARNING
The software has encountered an unusual non-fault condition.
.It Dv MM_INFO
The software informs about a non-error condition.
.El
.Pp
If no
.Fa severity
level is to be supplied,
.Dv MM_NOSEV
must be specified.
.Ss Text
The description of the condition the software encountered.
The character
string is not limited to a specific size.
.Pp
If no
.Fa text
is to be supplied,
.Dv MM_NOTXT
must be specified.
.Ss Action
The first step to be taken to recover from the condition the software
encountered; it will be preceded by the prefix
.Dq TO FIX: .
The character string is not limited to a specific size.
.Pp
If no
.Fa action
is to be supplied,
.Dv MM_NOACT
must be specified.
.Ss Tag
The on-line documentation which provides further information about the
condition and the message, such as
.Dq Xr fmtmsg 3 .
The character string is not limited to a specific size.
.Pp
If no
.Fa tag
is to be supplied,
.Dv MM_NOTAG
must be specified.
.Pp
Further effect on the formatting of the message as displayed on the
standard error stream (but not on the system console!) may be taken by
setting the
.Ev MSGVERB
environment variable, which selects the subset of message components
to be printed.
It consists of a colon-separated list of the optional keywords
.Fa label ,
.Fa severity ,
.Fa text ,
.Fa action ,
and
.Fa tag ,
which correspond to the arguments to
.Fn fmtmsg
with the same names.
If
.Ev MSGVERB
is either not set or malformed (containing empty or unknown keywords),
its content is ignored an all message components will be selected.
.Pp
Note that displaying a message on the system console may fail due to
inappropriate privileges or a non-permissive file mode of the console device.
.Sh RETURN VALUES
The
.Fn fmtmsg
function returns one of the following values:
.Bl -tag -width MM_NOTOKXXX
.It Dv MM_OK
The function succeeded.
.It Dv MM_NOTOK
The function failed completely.
.It Dv MM_NOMSG
The function was unable to generate a message on standard error,
but otherwise succeeded.
.It Dv MM_NOCOM
The function was unable to generate a message on the console,
but otherwise succeeded.
.El
.Sh SEE ALSO
.Xr printf 3 ,
.Xr syslog 3
.Sh STANDARDS
The
.Fn fmtmsg
function conforms to
.St -xsh5 .


482p3/lib/libc/gen/fmtmsg.c
/*	$NetBSD: fmtmsg.c,v 1.5 2012/03/20 16:36:05 matt Exp $	*/

/*-
 * Copyright (c) 1999 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Klaus Klein.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
__RCSID("$NetBSD: fmtmsg.c,v 1.5 2012/03/20 16:36:05 matt Exp $");
#endif /* LIBC_SCCS and not lint */

#include <fmtmsg.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static unsigned int	msgverb(const char *);
static const char *	severity2str(int);
static int		writeit(FILE *, unsigned int, const char *,
			    const char *, const char *, const char *,
			    const char *);

#define MM_VERBLABEL		0x01U
#define MM_VERBSEVERITY		0x02U
#define MM_VERBTEXT		0x04U
#define MM_VERBACTION		0x08U
#define MM_VERBTAG		0x10U
#define MM_VERBALL	\
    (MM_VERBLABEL | MM_VERBSEVERITY | MM_VERBTEXT | MM_VERBACTION | \
     MM_VERBTAG)

static const struct keyword {
	size_t			len;	/* strlen(keyword) */
	const char * const	keyword;
} keywords[] = {
	{ 5,	"label"		},	/* log2(MM_VERBLABEL) */
	{ 8,	"severity"	},	/* ... */
	{ 4,	"text"		},
	{ 6,	"action"	},
	{ 3,	"tag"		}	/* log2(MM_VERBTAG) */
};

static const size_t nkeywords = sizeof (keywords) / sizeof (keywords[0]);

/*
 * Convert a colon-separated list of known keywords to a set of MM_VERB*
 * flags, defaulting to `all' if not set, empty, or in presence of unknown
 * keywords.
 */
static unsigned int
msgverb(const char *str)
{
	u_int i;
	unsigned int result;

	if (str == NULL)
		return (MM_VERBALL);

	result = 0;
	while (*str != '\0') {
		for (i = 0; i < nkeywords; i++) {
			if (memcmp(str, keywords[i].keyword, keywords[i].len)
			    == 0 &&
			    (*(str + keywords[i].len) == ':' ||
			     *(str + keywords[i].len) == '\0'))
				break;
		}
		if (i == nkeywords) {
			result = MM_VERBALL;
			break;
		}

		result |= (1 << i);
		if (*(str += keywords[i].len) == ':')
			str++;	/* Advance */
	}
	if (result == 0)
		result = MM_VERBALL;

	return (result);
}

static const char severities[][8] = {
	"",		/* MM_NONE */
	"HALT",
	"ERROR",
	"WARNING",
	"INFO"
};

static const size_t nseverities = sizeof (severities) / sizeof (severities[0]);

/*
 * Returns the string representation associated with the numerical severity
 * value, defaulting to NULL for an unknown value.
 */
static const char *
severity2str(int severity)
{
	const char *result;

	if (severity >= 0 &&
	    (u_int) severity < nseverities)
		result = severities[severity];
	else
		result = NULL;

	return (result);
}

/*
 * Format and write the message to the given stream, selecting those
 * components displayed from msgverb, returning the number of characters
 * written, or a negative value in case of an error.
 */
static int
writeit(FILE *stream, unsigned int which, const char *label,
	const char *sevstr, const char *text, const char *action,
	const char *tag)
{
	int nwritten;

	nwritten = fprintf(stream, "%s%s%s%s%s%s%s%s%s%s%s",
	    ((which & MM_VERBLABEL) && label != MM_NULLLBL) ?
	    label : "",
	    ((which & MM_VERBLABEL) && label != MM_NULLLBL) ?
	    ": " : "",
	    (which & MM_VERBSEVERITY) ?
	    sevstr : "",
	    (which & MM_VERBSEVERITY) ?
	    ": " : "",
	    ((which & MM_VERBTEXT) && text != MM_NULLTXT) ?
	    text : "",
	    ((which & MM_VERBLABEL) && label != MM_NULLLBL) ||
	    ((which & MM_VERBSEVERITY)) ||
	    ((which & MM_VERBTEXT) && text != MM_NULLTXT) ?
	    "\n" : "",
	    ((which & MM_VERBACTION) && action != MM_NULLACT) ?
	    "TO FIX: " : "",
	    ((which & MM_VERBACTION) && action != MM_NULLACT) ?
	    action : "",
	    ((which & MM_VERBACTION) && label != MM_NULLACT) ?
	    " " : "",
	    ((which & MM_VERBTAG) && tag != MM_NULLTAG) ?
	    tag : "",
	    ((which & MM_VERBACTION) && action != MM_NULLACT) ||
	    ((which & MM_VERBTAG) && tag != MM_NULLTAG) ?
	    "\n" : "");

	return (nwritten);
}

int
fmtmsg(long classification, const char *label, int severity,
	const char *text, const char *action, const char *tag)
{
	FILE *console;
	const char *p, *sevstr;
	int result;

	/* Validate label constraints, if not null. */
	if (label != MM_NULLLBL) {
		/*
		 * Two fields, separated by a colon.  The first field is up to
		 * 10 bytes, the second is up to 14 bytes.
		 */
		p = strchr(label, ':');
		if (p ==  NULL || p - label > 10 || strlen(p + 1) > 14)
			return (MM_NOTOK);
	}
	/* Validate severity argument. */
	if ((sevstr = severity2str(severity)) == NULL)
		return (MM_NOTOK);

	/*
	 * Fact in search for a better place: XSH5 does not define any
	 * functionality for `classification' bits other than the display
	 * subclassification.
	 */

	result = 0;

	if (classification & MM_PRINT) {
		if (writeit(stderr, msgverb(getenv("MSGVERB")),
		    label, sevstr, text, action, tag) < 0)
			result |= MM_NOMSG;
	}
	/* Similar to MM_PRINT but ignoring $MSGVERB. */
	if (classification & MM_CONSOLE) {
		if ((console = fopen(_PATH_CONSOLE, "w")) != NULL) {
			if (writeit(console, MM_VERBALL,
			    label, sevstr, text, action, tag) < 0)
				result |= MM_NOCON;
			/*
			 * Ignore result: does not constitute ``generate a
			 * console message.''
			 */
			(void)fclose(console);
		} else {
			result |= MM_NOCON;
		}
	}

	if (result == (MM_NOMSG | MM_NOCON))
		result = MM_NOTOK;

	return (result == 0 ? MM_OK : result);
}
